### ISSUE TITLE
Dockerfile build doesn't support BuildKit

### ISSUE THREAD

**Issue by Tzrlk** [batect/batect#655]
When using the following batect container config:
```yaml
containers:
  workspace:
    build_directory: .
tasks:
  cmd:
    run:
      container: workspace
      command: ''
```
and the following Dockerfile:
```Dockerfile
# syntax=docker/dockerfile:1.1-experimental

FROM alpine:3.12

USER root:root
WORKDIR /root

# Install build tools
RUN --mount=type=cache,id=apk,target=/etc/apk/cache,sharing=locked \
	apk add \
		curl~=7.67
```
I receive the following error:
```
>batect cmd
Running cmd...
workspace: building image...

Error: Could not build image for container workspace.
Building image failed: Dockerfile parse error line 9: Unknown flag: mount

Clean up: done

The task cmd failed. See above for details.
```
I've got BuildKit enabled in both environment variables:
```
>env | grep BUILDKIT
DOCKER_BUILDKIT=1
```
and in the daemon config:
```json
{
  "debug": true,
  "experimental": true,
  "features": {
    "buildkit": true
  }
}
```
And regular docker builds succeed without issues:
```
>docker build .
[+] Building 1.9s (9/9) FINISHED
 => [internal] load .dockerignore                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                 0.0s
 => [internal] load build definition from Dockerfile                                                                                            0.0s
 => => transferring dockerfile: 2.50kB                                                                                                          0.0s
 => resolve image config for docker.io/docker/dockerfile:1.1-experimental                                                                       0.9s
 => CACHED docker-image://docker.io/docker/dockerfile:1.1-experimental@sha256:8c69d118cfcd040a222bea7f7d57c6156faa938cb61b47657cd65343babc3664  0.0s
 => [internal] load metadata for docker.io/library/alpine:3.12                                                                                  0.9s
 => [stage-0 1/3] FROM docker.io/library/alpine:3.12@sha256:a15790640a6690aa1730c38cf0a440e2aa44aaca9b0e8931a9f2b0d7cc90fd65                    0.0s
 => CACHED [stage-0 2/3] WORKDIR /root                                                                                                          0.0s
 => CACHED [stage-0 3/3] RUN --mount=type=cache,id=apk,target=/etc/apk/cache,sharing=locked  apk add   curl                                     0.0s
 => exporting to image                                                                                                                          0.0s
 => => exporting layers                                                                                                                         0.0s
 => => writing image sha256:a59947283439e1f9cd073ebdcd265be4ba7007df7a279ae40251fcdaea802550                                                    0.0s
```
Why is batect ignoring the BuildKit feature flag? I would have thought setting it in the daemon config would have ensured its use (though IntelliJ seems to have the same problem).

**Comment by charleskorn** [comment]
Hey @Tzrlk, thanks for the detailed issue report. batect doesn't currently support BuildKit, but it's been on my to-do list for a while. I can't promise when it will be available, but I'll take a look at it soon.

To your question about the `DOCKER_BUILDKIT` environment variable: this only applies if the Docker CLI is used - batect uses the Docker API directly, and I'm guessing IntelliJ does the same, which is why you're seeing that be ignored.

**Comment by Tzrlk** [comment]
I figured it'd be something like that. Strangest is that the daemon configuration for BuildKit doesn't flow-through to the api. Not like strange design decisions aren't par for the course with Docker, though.

**Comment by charleskorn** [comment]
I've just released v0.60 which has experimental support for BuildKit. Please give it a go and let me know if you run into any issues.

**Comment by Tzrlk** [comment]
Appears to switch to the BuildKit engine just fine, though has problems pulling the syntax image from either a private:
```
Running terragrunt:cluster...
workspace: building image: resolve image config for <private_repo>/docker/dockerfile:1.1-experimental

Error: Could not build image for container workspace.
Building image failed: failed to solve with frontend dockerfile.v0: failed to solve with frontend gateway.v0: failed to fetch anonymous token: unexpected status: 401 Unauthorized. Output from build process was:
#1 [internal] load remote build context
#1 CACHED

#2 copy /context /
#2 CACHED

#3 resolve image config for <private_repo>/docker/dockerfile:1.1-experimental
#3 ERROR: failed to fetch anonymous token: unexpected status: 401 Unauthorized

Clean up: done

The task terragrunt:cluster failed. See above for details.
```
or public repo:
```
Running terragrunt:cluster...
workspace: building image: resolve image config for docker.io/docker/dockerfile:1.1-experimental

Error: Could not build image for container workspace.
Building image failed: failed to solve with frontend dockerfile.v0: failed to solve with frontend gateway.v0: failed to fetch anonymous token: unexpected status: 401 Unauthorized. Output from build process was:
#1 [internal] load remote build context
#1 DONE

#2 copy /context /
#2 DONE

#3 resolve image config for docker.io/docker/dockerfile:1.1-experimental
#3 ERROR: failed to fetch anonymous token: unexpected status: 401 Unauthorized

Clean up: done

The task terragrunt:cluster failed. See above for details.
```
If I remove the syntax image reference (and any features that depend on it), the build seems to run fine, though with way less logging (as BuildKit is want to do without further configuration).

**Comment by charleskorn** [comment]
OK, that makes sense - BuildKit handles registry authentication slightly differently and it looks like batect isn't quite doing the right thing.

Does using an image from a private repo that requires authentication work for you?

**Comment by Tzrlk** [comment]
> Does using an image from a private repo that requires authentication work for you?

Regularly, yeah, and even with regular BuildKit builds. Just need to `docker login <registry>` at some point beforehand.
Only seems to have trouble with the syntax image.

**Comment by charleskorn** [comment]
Thanks for the info - I can't promise when I'll be able to fix this, but it's high on my priority list. I'm waiting for https://github.com/square/wire/issues/1325 to be resolved before I can pick this up.

**Comment by charleskorn** [comment]
🚨 Batect's experimental support for BuildKit does not work with Docker 20.10. I'll take a look at this in the next few days to see if it's an easy fix, but at first glance it looks like it's related to the changes that are blocked by that Wire issue linked above.

**Comment by charleskorn** [comment]
I've just released 0.64.2 with a bunch of improvements to BuildKit support, including a fix for Docker 20.10 and support for pulling images from registries that require credentials. 

Based on my testing, this should fix all of the issues discussed in this issue, so I'm going to go ahead and close it. If that's not the case, please feel free to comment here or open a new issue if you run into any other problems.
